src/concepts/functors.md

```diff
@@ -0,0 +1,649 @@
+                      end) :> STACK =
+struct
+    type 'a t = 'a S.t
+
+    val limit = 10
+    exception Full
+
+    fun push S x =
+        if S.size S >= limit
+          then raise Full
+          else S.push S x
+
+    fun pop S = S.pop S
+
+    fun size S = S.size S
+
+    fun empty S = S.empty
+
+end
+
+structure BoundedStack1 = BoundedStack(Stack1)
+structure BoundedStack2 = BoundedStack(Stack2)
+```
+
+Thus, `BoundedStack` is no longer a functor taking in a structure ascribing to
+`STACK`, but a functor taking in a structure ascribing to `sig structure S:
+STACK end`. In other words, taking in a structure _containing_ a structure
+ascrbing to `STACK`! Thus, the line `structure BoundedStack1 =
+BoundedStack(Stack1)` will not type-check, as `Stack1` does not ascribe to the
+same signature that `BoundedStack` is expecting. This one simple syntax error
+can be the source of much pain and frustration, so we caution the reader to be
+particular with their syntax, and mindful of what is really happening under the
+hood.
+
+## Case study: Typeclasses
+
+Certain types have some functionality or operation in common. Depending on the
+operation in question, we can say that these types fall into the same
+_typeclass_, which is a common interface consisting of a type and the desired
+operations. Note that typeclass membership is not a formally defined
+relationship, but instead a useful categorization that we use in order to
+classify types that we intend to parameterize some implementation over.
+
+For a concrete example of a typeclass, consider the `ORDERED` typeclass.
+
+```sml
+signature ORDERED =
+sig
+    type t
+    val compare : t * t -> order
+end
+```
+
+The `ORDERED` typeclass consists of those types that admit a _sensible
+ordering_, which we will not (and perhaps cannot) define. Thus, we can witness
+`int` and `string` as valid instances of the `ORDERED` typeclass with the
+following structures:
+
+```sml
+structure IntOrder : ORDERED =
+struct
+    type t = int
+    val compare = Int.compare
+end
+
+structure StringOrder : ORDERED =
+struct
+    type t = string
+    val compare = String.compare
+end
+```
+
+Note that it is useful, in this case, for our instances of the `ORDERED`
+typeclass to be _transparently ascribed_, since it is the whole point that we
+are aware of the type that the typeclass is associated with.
+
+These definitions come very naturally from the fact that the `String` and `Int`
+libraries included with the Standard ML basis library already implement
+`compare` fields, however we can also define types such as `int list` to be
+instances of `ORDERED`:
+
+```sml
+structure IntListOrder : ORDERED =
+struct
+    type t = int list
+    fun compare [] [] = EQUAL
+      | compare [] ys = LESS
+      | compare xs [] = GREATER
+      | compare (x::xs) (y::ys) =
+        case Int.compare (x, y) of
+            EQUAL => compare xs ys
+          | LESS => LESS
+          | GREATER => GREATER
```

Comment:
Also this shouldn't be curried, so it can even just be `val compare = List.collate Int.compare`

Reply:
Do we commonly teach `List.collate`? If not, it might be worth explaining its function elsewhere (and perhaps using exactly this `compare` function as an example) and linking to it.